### PR TITLE
Fixes #15011: Error at the end of a policy generation with too many nodes 

### PR DESCRIPTION
**Pull request by fanf**
https://issues.rudder.io/issues/15011

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala [hunk 1]
```diff
@@ -1439,3 +1421,130 @@ trait PromiseGeneration_historization extends PromiseGenerationService {
...
+             if(file.exists) {
+               file.moveTo(savedOld, true)
+             }
+           }(s"Can not move previous updated node IDs file to '${savedOld.pathAsString}'")
+      _ <- effect {
+            // header
+            file.writeText(s"# This file contains IDs of nodes updated by policy generation started at '${date(start)}' ended at '${date(end)}'\n\n")
+            // policy servers
+            file.writeText(s"\nRUDDER_UPDATED_POLICY_SERVER_IDS=${formatIds(policyServers)}\n")
+            file.writeText(s"\nRUDDER_UPDATED_SIMPLE_NODE_IDS=${formatIds(simpleNodes)}\n")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala]: i would not use SIMPLE_NODE, but simply NODE

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/resources/hooks.d/policy-generation-finished/60-trigger-node-update [hunk 2]
```diff
@@ -57,7 +58,7 @@ fi
 if [ -n "${RUDDER_NODE_IDS}" ] && (( ${TO_UPDATE} > 0 )) ; then
 
   # split nodes in an array
-  read -ra NODES_ARRAY <<<"${RUDDER_NODE_IDS}"
+  read -ra NODES_ARRAY <<<$(echo "${RUDDER_NODE_IDS}" | tr '\n' ' ')
   # takes only the TO_UPDATE first nodes - nodes id are expected to be sorted as wanted
   NODES=$(IFS=, eval 'echo "${NODES_ARRAY[*]:0:${TO_UPDATE}}"')
```
Comment (peckpeck) [inline webapp/sources/rudder/rudder-core/src/main/resources/hooks.d/policy-generation-finished/60-trigger-node-update]: replace those 2 by :
NODES_LIST="${NODES_ARRAY[*]:0:${TO_UPDATE}}"
NODES="${NODES_LIST// /,}"

**Comment by fanf** [comment]
PR updated with a new commit

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala [hunk 3]
```diff
@@ -318,11 +326,8 @@ trait PromiseGenerationService {
       timeComputeGroups    =  (System.currentTimeMillis - initialTime)
       _                    =  PolicyLogger.debug(s"Computing dynamic groups finished in ${timeComputeGroups} ms")
       preGenHooksTime      =  System.currentTimeMillis
-
-      //fetch all
-      preHooks             <- RunHooks.getHooks(HOOKS_D + "/policy-generation-started", HOOKS_IGNORE_SUFFIXES)
-      _                    <- RunHooks.syncRun(preHooks, HookEnvPairs.build( ("RUDDER_GENERATION_DATETIME", generationTime.toString) ), systemEnv)
       timeRunPreGenHooks   =  (System.currentTimeMillis - preGenHooksTime)
+      _                    <- runPreHooks(generationTime, systemEnv)
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala]: this hsould be before the timeRunPreGenHooks

**Comment by fanf** [comment]
PR updated with a new commit

**Comment by fanf** [comment]
OK, squash merging this PR

**Pull request merged**